### PR TITLE
Add verification of image manifest digests on v2 push/pull

### PR DESCRIPTION
**Pull request by jlhawn**
~~Depends on https://github.com/docker/docker/pull/11271~~

### CODE REVIEW

**Comment by jlhawn** [comment]
ping @dmcgowan @stevvooe @ncdc

**Comment by jlhawn** [comment]
@jfrazelle can we add this to the 1.6 milestone? It complements @ncdc's pull by digest change. #11271 introduces the distribution/digest library which makes the manifest digest verification relatively simple in this PR. I'll rebase once that's merged.

**Comment by jfrazelle** [comment]
added

**Comment by jlhawn** [comment]
@jfrazelle @tiborvass @icecrime @dmcgowan @ncdc #11271 has been merged (hooray!) so I've rebased this onto master. Jenkins is currently running but it should pass (barring the current daemon crash on jenkins issue).  Please review!

**Comment by jlhawn** [comment]
cc @LK4D4 @crosbymichael @vbatts @stevvooe and anyone who's available at this point, really.

File: registry/session_v2.go [hunk 1]
```diff
@@ -95,11 +98,43 @@ func (r *Session) GetV2ImageManifest(ep *Endpoint, imageName, tagName string, au
...
+
+	dgstVerifier, err := digest.NewDigestVerifier(hdrDigest)
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to verify manifest digest from registry: %s", err)
+	}
+
+	dgstVerifier.Write(payload)
+
+	if !dgstVerifier.Verified() {
+		computedDigest, _ := digest.FromBytes(payload)
```
Comment (ncdc) [inline registry/session_v2.go]: Is there any risk that the Daemon would compute a different digest than the registry?

@stevvooe @dmcgowan @jlhawn

Comment (jlhawn) [inline registry/session_v2.go]: If the registry has gone rogue, then maybe.

Comment (ncdc) [inline registry/session_v2.go]: Is the payload guaranteed to be the same as what the registry is storing?
Any chance the fields could show up in a different order?

Comment (jlhawn) [inline registry/session_v2.go]: It should be unaltered from what was pushed, yes. What's pushed is a manifest payload with attached signatures and the digest is only of the payload.

Comment (ncdc) [inline registry/session_v2.go]: Ok, not being familiar with the signature bits, I wasn't sure if it was
getting marshaled and unmarshaled.

**Comment by dmp42** [comment]
@jlhawn build is failing

**Comment by ncdc** [comment]
@dmp42 unrelated, it's the sporadic TestPullVerified failure

**Comment by jlhawn** [comment]
@ncdc @dmp42 That test was failing because our mock-v2 registry which serves official images to docker 1.5 doesn't include the Digest header so the digest was always invalid and it would always fall back to v1 pulls.

I've just updated it to only verify the digest header *if* it's actually pulling by digest reference.

File: registry/session_v2.go [hunk 2]
```diff
@@ -95,11 +98,47 @@ func (r *Session) GetV2ImageManifest(ep *Endpoint, imageName, tagName string, au
...
 	}
 
-	buf, err := ioutil.ReadAll(res.Body)
+	manifestBytes, err := ioutil.ReadAll(res.Body)
 	if err != nil {
 		return nil, "", fmt.Errorf("Error while reading the http response: %s", err)
 	}
-	return buf, res.Header.Get(DockerDigestHeader), nil
+
+	hdrDigest := digest.Digest(res.Header.Get(DockerDigestHeader))
```
Comment (stevvooe) [inline registry/session_v2.go]: Use `digest.ParseDigest` for this.

File: registry/session_v2.go [hunk 3]
```diff
@@ -95,11 +98,47 @@ func (r *Session) GetV2ImageManifest(ep *Endpoint, imageName, tagName string, au
...
+	if res.Header.Get(DockerDigestHeader) != "" {
+		// Need to verify the digest.
+
+		sig, err := libtrust.ParsePrettySignature(manifestBytes, "signatures")
```
Comment (stevvooe) [inline registry/session_v2.go]: Can we move this to a separate function? I think we can replace this with something from distribution and that will be easier if there is an existing function.

**Comment by jlhawn** [comment]
@stevvooe I've moved the digest verification on pull to the same function which already verifies the manifest signature. PTAL.

File: graph/manifest.go [hunk 4]
```diff
@@ -32,6 +34,31 @@ func (s *TagStore) loadManifest(eng *engine.Engine, manifestBytes []byte) (*regi
...
+
+		dgstVerifier.Write(payload)
+
+		if !dgstVerifier.Verified() {
+			computedDigest, _ := digest.FromBytes(payload)
+			return nil, false, fmt.Errorf("unable to verify manifest diges: registry has %q, computed %q", manifestDigest, computedDigest)
+		}
+	}
+
+	if utils.DigestReference(ref) && ref != manifestDigest.String() {
```
Comment (stevvooe) [inline graph/manifest.go]: I'm confused as to why `ref` and `dgst` are passed in. Isn't a ref either a tag or digest?

Comment (jlhawn) [inline graph/manifest.go]: `dgst` is the digest from the registry response header (only a string value for now, it's later parsed to `manifestDigest`), verification of this value is deferred to this function.

`ref` is what the user used to pull the image. If it's a digest reference, then it must match the verified `manifestDigest`.

File: registry/session_v2.go [hunk 5]
```diff
@@ -297,7 +300,24 @@ func (r *Session) PutV2ImageManifest(ep *Endpoint, imageName, tagName string, ma
 		return "", utils.NewHTTPRequestError(fmt.Sprintf("Server error: %d trying to push %s:%s manifest", res.StatusCode, imageName, tagName), res)
 	}
 
-	return res.Header.Get(DockerDigestHeader), nil
+	hdrDigest, err := digest.ParseDigest(res.Header.Get(DockerDigestHeader))
```
Comment (stevvooe) [inline registry/session_v2.go]: Is this section verifying the digest received after PUTing the manifest?

File: graph/manifest.go (hunk 4 above)
Comment (stevvooe) [inline graph/manifest.go]: Okay, that is a nice distinction. We should follow that convention.

**Comment by stevvooe** [comment]
LGTM.

A few questions but nothing to hold up the show.

**Comment by dmp42** [comment]
LGTM

File: registry/session_v2.go (hunk 5 above)
Comment (jlhawn) [inline registry/session_v2.go]: yep.

Comment (jlhawn) [inline registry/session_v2.go]: then makes sure that docker got the same digest as the registry.

**Comment by jlhawn** [comment]
@ncdc @jfrazelle PTAL

File: graph/manifest.go [hunk 6]
```diff
@@ -32,6 +34,31 @@ func (s *TagStore) loadManifest(eng *engine.Engine, manifestBytes []byte) (*regi
...
+		dgstVerifier, err := digest.NewDigestVerifier(manifestDigest)
+		if err != nil {
+			return nil, false, fmt.Errorf("unable to verify manifest digest from registry: %s", err)
+		}
+
+		dgstVerifier.Write(payload)
+
+		if !dgstVerifier.Verified() {
+			computedDigest, _ := digest.FromBytes(payload)
+			return nil, false, fmt.Errorf("unable to verify manifest diges: registry has %q, computed %q", manifestDigest, computedDigest)
```
Comment (dmcgowan) [inline graph/manifest.go]: typo "diges"

**Comment by dmcgowan** [comment]
Tested push, verified manifest, and tampered manifest

LGTM

**Comment by ncdc** [comment]
It's late for me - I'll review first thing tomorrow.

File: graph/pull.go [hunk 7]
```diff
@@ -430,12 +430,15 @@ func (s *TagStore) pullV2Repository(eng *engine.Engine, r *registry.Session, out
 
 func (s *TagStore) pullV2Tag(eng *engine.Engine, r *registry.Session, out io.Writer, endpoint *registry.Endpoint, repoInfo *registry.RepositoryInfo, tag string, sf *utils.StreamFormatter, parallel bool, auth *registry.RequestAuthorization) (bool, error) {
 	log.Debugf("Pulling tag from V2 registry: %q", tag)
+
+	// GetV2ImageManifest ensures that the manifest content has the expected
+	// digest if the tag is a digest reference.
```
Comment (ncdc) [inline graph/pull.go]: Update/move comment to reflect that this verification is now in `loadManifest`?

**Comment by ncdc** [comment]
1 comment on a comment, otherwise LGTM

**Comment by jlhawn** [comment]
@ncdc @dmcgowan I've fixed the typo/comments and rebased on upstream/master. :+1:

**Comment by ncdc** [comment]
LGTM

**Comment by dmcgowan** [comment]
LGTM

**Comment by jfrazelle** [comment]
LGTM

**Comment by jlhawn** [comment]
:birthday:

**Pull request merged**